lib/github.ts

```diff
@@ -9,44 +9,73 @@ export const getLatestReleaseForBranch = async (
     branch: string;
   },
 ): Promise<GitHubRelease | null> => {
+  // the goal of this function is to find the latest release for a given branch. 
+  // the github api does not provide an easy way to do this. you must get releases and commits separately and then compare them.
+  //
+  // both of these endpoints have paging and we also want to minimize the number of requests we make to github for rate limiting. 
+  // 
+  // the strategy here is a mix between performance and simplicity for maintenance of the code. 
+  // First, get all releases for the repo. Even if we don't need to view them all, get them all otherwise we may need to implement some complex caching strategy. 
+  // Next, page through commits for a branch and compare to all releases. 
+  // We are assuming there are less releases than commits, which is what makes this performant. 
```

Comment:
I was noticing that the tool was performing a lot of api requests even for small repositories without a lot of releases or commits. It was due to a bug where we perform a api fetch for releases for every single commit. 

By changing the logic of this function, it is much more performant. We test this logic via automated tests.